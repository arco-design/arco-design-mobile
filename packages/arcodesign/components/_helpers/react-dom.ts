import { ReactElement } from 'react';
import * as ReactDOM from 'react-dom';

function isObject(obj: any): obj is { [key: string]: any } {
    return Object.prototype.toString.call(obj) === '[object Object]';
}

export interface RootType {
    render: (element: ReactElement) => void;
    _unmount: () => void;
}
export interface RootTypeReact extends RootType {
    unmount?: () => void;
}
export type CreateRootFnType = (container: Element | DocumentFragment) => {
    render: (element: ReactElement) => void;
    unmount?: () => void;
};

// Cast ReactDOM to a version that might have createRoot and other properties
const typedReactDOM = ReactDOM as typeof ReactDOM & {
    createRoot?: CreateRootFnType;
    render?: (
        element: ReactElement,
        container: Element | DocumentFragment,
        callback?: () => void,
    ) => any;
    unmountComponentAtNode?: (container: Element | DocumentFragment) => boolean;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
};

let copyRender: (
    app: ReactElement,
    container: Element | DocumentFragment,
    createRootFunction?: CreateRootFnType,
) => RootType;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
    // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
    // Avoid console warning
    const secretInternals = typedReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    if (isObject(secretInternals)) {
        secretInternals.usingClientEntryPoint = skipWarning;
    }
};

const createRootFn = typedReactDOM.createRoot;

const getRender =
    (createRootFunction: CreateRootFnType) =>
    (app: ReactElement, container: Element | DocumentFragment) => {
        updateUsingClientEntryPoint(true);
        const root = createRootFunction!(container);
        updateUsingClientEntryPoint(false);

        root.render(app);

        return {
            render: (elementToRender: ReactElement) => {
                root.render(elementToRender);
            },
            _unmount() {
                setTimeout(() => {
                    if (root && typeof root.unmount === 'function') {
                        root.unmount();
                    }
                });
            },
        };
    };

if (createRootFn) {
    // React 18
    copyRender = getRender(createRootFn);
} else if (
    typeof typedReactDOM.render === 'function' &&
    typeof typedReactDOM.unmountComponentAtNode === 'function'
) {
    // React 16/17
    copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
        typedReactDOM.render!(app, container); // Use non-null assertion
        return {
            render: (elementToRender: ReactElement) => {
                typedReactDOM.render!(elementToRender, container); // Use non-null assertion
            },
            _unmount() {
                typedReactDOM.unmountComponentAtNode!(container); // Use non-null assertion
            },
        };
    };
} else {
    copyRender = (
        app: ReactElement,
        container: Element | DocumentFragment,
        createRootFunction?: CreateRootFnType,
    ) => {
        const defaultCb = () => {
            // Fallback if no rendering method is found
            console.error(
                'ArcoDesign: ReactDOM.createRoot() or ReactDOM.render() not found. ' +
                    'This usually means you are using an unsupported version of React, ' +
                    'or ReactDOM is not properly initialized. ' +
                    'ArcoDesign Mobile React requires React 16, 17, 18, or 19.',
            );
            return { render: (_element: ReactElement) => {}, _unmount: () => {} };
        };
        if (createRootFunction) {
            return getRender(createRootFunction)(app, container);
        }
        return defaultCb();
    };
}

export const render = copyRender;
