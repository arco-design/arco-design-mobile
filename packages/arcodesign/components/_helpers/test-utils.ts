import React, { useEffect, useRef } from 'react';
import { Promise as ES6Promise, Promise } from 'es6-promise';

/**
 * React 19 兼容的 act 包装器
 * 用于包装可能触发状态更新的异步操作
 */
export const actWrapper = (callback: () => void | ES6Promise<void>): ES6Promise<void> => {
    return new ES6Promise<void>((resolve, reject) => {
        try {
            // 尝试使用 react-dom/test-utils 的 act
            const testUtils = require('react-dom/test-utils');
            if (testUtils && typeof testUtils.act === 'function') {
                // 对于同步回调
                if (typeof callback === 'function') {
                    try {
                        const result = callback();
                        if (result && typeof result.then === 'function') {
                            testUtils
                                .act(() => result)
                                .then(() => resolve())
                                .catch(reject);
                        } else {
                            testUtils.act(() => {
                                callback();
                            });
                            resolve();
                        }
                    } catch (error) {
                        reject(error);
                    }
                } else {
                    resolve();
                }
            } else {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            }
        } catch (error) {
            // 如果 act 不可用，直接执行回调
            try {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            } catch (callbackError) {
                reject(callbackError);
            }
        }
    });
};

/**
 * 清理未处理的异步操作，防止内存泄漏警告
 */
export const cleanupAsync = (): ES6Promise<void> => {
    return new ES6Promise<void>(resolve => {
        setTimeout(() => resolve(), 0);
    });
};

/**
 * 安全的状态更新，检查组件是否已卸载
 */
export const createSafeStateUpdater = (isMountedRef: React.MutableRefObject<boolean>) => {
    return (updater: () => void) => {
        if (isMountedRef.current) {
            updater();
        }
    };
};

/**
 * 用于测试的 useEffect 清理函数
 */
export const useTestCleanup = (cleanupFn: () => void) => {
    const isMountedRef = useRef(true);

    useEffect(() => {
        return () => {
            isMountedRef.current = false;
            cleanupFn();
        };
    }, [cleanupFn]);

    return isMountedRef;
};
