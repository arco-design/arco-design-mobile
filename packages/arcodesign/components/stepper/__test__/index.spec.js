import React, { createRef } from 'react';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import { defaultContext } from '../../context-provider';
import Stepper from '..';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

demoTest('stepper');

mountTest(Stepper, 'Stepper');

const prefix = `.${defaultContext.prefixCls}-stepper`;
const minusButton = `${prefix}-minus-button`;
const input = `${prefix}-input`;
const addButton = `${prefix}-add-button`;

describe('Stepper', () => {
    // 组件正常渲染
    it('component render correctly', () => {
        const component = mount(<Stepper />);
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        // 正常渲染
        expect(minusButtonDom.length).toBe(1);
        expect(inputDom.length).toBe(1);
        expect(addButtonDom.length).toBe(1);
        expect(component.find(prefix).length).toBe(1);

        // 默认值为1
        expect(inputDom.instance().value).toEqual('1');
    });

    // 按钮点击数据结果正常
    it('button click correctly', () => {
        const component = mount(<Stepper defaultValue={1} min={0} max={20} step={1} />);
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        // 点击减号
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0');

        // 点击加号
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1');
    });

    // 数据边界正确
    it('data limit correctly', () => {
        const component = mount(<Stepper defaultValue={1} min={0} max={1} step={1} />);
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        // 最小值为0
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0');
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0');

        // 最大值为1
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1');
    });

    // 传入受控值
    it('accept value', () => {
        const component = mount(<Stepper value={1} min={0} max={100} step={1} />);
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        // 值受控，点击加减不触发数据变动
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1');
    });

    // 内容是否允许为空
    it('input value allow empty', () => {
        // 允许为空
        const component = mount(<Stepper defaultValue={1} min={0} max={100} step={1} allowEmpty />);
        const inputDom = component.find(input);

        inputDom.instance().value = '';
        inputDom.simulate('blur');
        expect(inputDom.instance().value).toEqual('');

        // 不允许为空
        const component1 = mount(<Stepper defaultValue={1} min={0} max={100} step={1} />);
        const inputDom1 = component1.find(input);

        inputDom1.instance().value = '';
        inputDom1.simulate('blur');
        expect(inputDom1.instance().value).toEqual('1');

        inputDom1.instance().value = '101';
        inputDom1.simulate('blur');
        expect(inputDom1.instance().value).toEqual('100');

        inputDom1.instance().value = '-1';
        inputDom1.simulate('blur');
        expect(inputDom1.instance().value).toEqual('0');
    });

    // 小数点
    it('data digits', () => {
        const component = mount(
            <Stepper defaultValue={0} min={-1} max={100} step={0.5} digits={1} />,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        expect(inputDom.instance().value).toEqual('0.0');

        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-0.5');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.5');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1.0');
    });

    // 禁用步进器
    it('stepper disable', () => {
        const component = mount(
            <Stepper defaultValue={0} min={-1} max={100} step={0.5} digits={1} disabled />,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
    });

    // 输入框只读
    it('stepper input readonly', () => {
        const component = mount(
            <Stepper defaultValue={0} min={-1} max={100} step={0.5} digits={1} inputReadonly />,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        expect(inputDom.instance().value).toEqual('0.0');

        inputDom.simulate('change', {
            target: {
                value: '2.0',
            },
        });
        expect(inputDom.instance().value).toEqual('0.0');

        // 按钮点击行为正常
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-0.5');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.5');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1.0');
    });

    // 不允许操作大于/小于极限值时，操作结果等于极限值
    it('equal limit disable', () => {
        const component = mount(
            <Stepper defaultValue={0} min={-1} max={1} step={0.7} digits={1} equalLimitDisabled />,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-0.7');
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-0.7');

        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.0');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.7');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.7');
    });

    // 允许操作大于/小于极限值时，操作结果等于极限值
    it('equal limit disable', () => {
        const component = mount(
            <Stepper defaultValue={0} min={-1} max={1} step={0.7} digits={1} />,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-0.7');
        minusButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-1.0');

        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('-0.3');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('0.4');
        addButtonDom.simulate('click');
        expect(inputDom.instance().value).toEqual('1.0');
    });

    // 格式化当前值
    it('formatter current value', done => {
        const formatter = currentValue =>
            new Promise(resolve => {
                setTimeout(() => {
                    resolve(currentValue + 1);
                }, 200);
            });

        const component = mount(
            <Stepper defaultValue={0} min={-1} max={5} step={1} digits={3} formatter={formatter} />,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const delay = 300;
        minusButtonDom.simulate('click');
        setTimeout(() => {
            component.update();
            expect(inputDom.instance().value).toEqual('0.000');
            done();
        }, delay);
    });

    // 用户自定义按钮和输入框
    it('render by myself', () => {
        const minusButton = <div className="my-minus-button"></div>;
        const addButton = <div className="my-add-button"></div>;
        const component = mount(
            <Stepper
                defaultValue={0}
                min={-1}
                max={5}
                step={1}
                addButton={addButton}
                minusButton={minusButton}
                renderContent={currentValue => {
                    const newValue = currentValue + 10;
                    return <input className="my-input" value={newValue} type="number" />;
                }}
            />,
        );
        const minusButtonDom = component.find('.my-minus-button');
        const inputDom = component.find('.my-input');
        const addButtonDom = component.find('.my-add-button');

        expect(minusButtonDom.length).toBe(1);
        expect(inputDom.length).toBe(1);
        expect(addButtonDom.length).toBe(1);
        expect(inputDom.instance().value).toEqual('10');
    });

    // change empty
    it('change value', done => {
        const ref = createRef();
        const formatter = currentValue => currentValue + 1;
        const component = mount(
            <Stepper ref={ref} formatter={formatter} defaultValue={0} min={-1} max={1} />,
        );
        const { changeValue } = ref.current;
        act(() => {
            changeValue();
        });
        component.update();
        setTimeout(() => {
            expect(component.find(input).instance().value).toEqual('1');
            done();
        }, 500);
    });

    // change value max
    it('change value1', done => {
        const ref = createRef();
        const formatter = currentValue => currentValue + 1;
        const component = mount(
            <Stepper ref={ref} formatter={formatter} defaultValue={0} min={-1} max={1} />,
        );
        const { changeValue } = ref.current;
        // jest.mock(changeValue);
        act(() => {
            changeValue(2);
        });
        component.update();
        setTimeout(() => {
            expect(component.find(input).instance().value).toEqual('1');
            done();
        }, 500);
    });

    // change value min
    it('change value2', done => {
        const ref = createRef();
        const formatter = currentValue => currentValue - 1;
        const component = mount(
            <Stepper ref={ref} formatter={formatter} defaultValue={0} min={-1} max={1} />,
        );

        act(() => {
            ref.current.changeValue(-2);
        });
        component.update();
        setTimeout(() => {
            expect(component.find(input).instance().value).toEqual('-1');
            done();
        }, 200);
    });

    // 输入框 input 事件
    it('stepper input event', () => {
        const component = mount(
            <Stepper defaultValue={0} min={-1} max={100} step={0.5} digits={1} allowEmpty/>,
        );
        const minusButtonDom = component.find(minusButton);
        const inputDom = component.find(input);
        const addButtonDom = component.find(addButton);

        expect(inputDom.instance().value).toEqual('0.0');

        inputDom.simulate('input', {
            target: {
                value: '2.2',
            },
        });
        expect(inputDom.instance().value).toEqual('2.2');

        inputDom.simulate('input', {
            target: {
                value: '2.222222',
            },
        });
        expect(inputDom.instance().value).toEqual('2.2');


        inputDom.simulate('input', {
            target: {
                value: '',
            },
        });
        expect(inputDom.instance().value).toEqual('');

        inputDom.simulate('input', {
            target: {
                value: '44',
            },
        });
        expect(inputDom.instance().value).toEqual('44.0');
    });
});
