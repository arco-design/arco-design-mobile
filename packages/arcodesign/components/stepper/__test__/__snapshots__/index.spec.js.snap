// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stepper demo test stepper demo: default.md renders correctly 1`] = `
<div
  class="arco-cell-group all-border-box"
>
  <div
    class="cell-group-body"
  >
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            单边禁用
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            可用状态
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1.0"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            禁用状态
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`stepper demo test stepper demo: events.md renders correctly 1`] = `
<div
  class="arco-cell-group all-border-box"
>
  <div
    class="cell-group-body"
  >
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            onBlur
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            onChange
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            onFocus
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            onAddButtonClick
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            onMinusButtonClick
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            onClick
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`stepper demo test stepper demo: renderContent.md renders correctly 1`] = `
<div
  class="arco-cell-group all-border-box"
>
  <div
    class="cell-group-body"
  >
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            addButton
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div>
                add
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            minusButton
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div>
                minus
              </div>
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            renderContent
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              type="number"
              value="11"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`stepper demo test stepper demo: style.md renders correctly 1`] = `
<div
  class="arco-cell-group all-border-box"
>
  <div
    class="cell-group-body"
  >
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            只读
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button"
              />
            </div>
            <input
              class="arco-stepper-input"
              readonly=""
              type="number"
              value="5"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            显示边框
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-square"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            圆角风格
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-round"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            style
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
            style="background:red"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            className
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default wxy-test"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            inputStyle
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              style="background:blue"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            inputClass
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input wxy-test"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`stepper demo test stepper demo: value.md renders correctly 1`] = `
<div
  class="arco-cell-group all-border-box"
>
  <div
    class="cell-group-body"
  >
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            设置最大/最小值
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="11"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            设置小数点
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="10.0"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            设置递增值
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            允许空值
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            可等于极限
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1.3"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            formatter
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1.0"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-cell all-border-box bordered"
    >
      <div
        class="arco-cell-inner"
      >
        <div
          class="cell-label"
        >
          <div
            class="cell-title"
          >
            change value
          </div>
        </div>
        <div
          class="cell-content has-label"
        >
          <div
            class="arco-stepper arco-default"
          >
            <div>
              <div
                class="arco-stepper-minus-button disabled"
              />
            </div>
            <input
              class="arco-stepper-input"
              type="number"
              value="1"
            />
            <div>
              <div
                class="arco-stepper-add-button"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
